File: 0x02-functions_nested_loops/fib.js
#!/usr/bin/node
const process = require('node:process');

function createAddClosure () {
	/** declare and initialize globSum, encapsulated in closure creator func **/
	let globSum;

	/** define add function, returned by the closure creator func **/
	function add(a, b) {
		let lenA = a.length;
		let lenB = b.length;
		let i = lenA - 1;
		let j = lenB - 1;
		let k = 0;
		let digitA;
		let digitB;
		let sum;
		let carry = 0;
		let start;
		let end;
		let temp;
		let tempSum = '';

		while (i >= 0 || j >= 0 || carry) {
			digitA = (i >= 0) ? parseInt(a[i]) : 0;
			digitB = (j >= 0) ? parseInt(b[j]) : 0;
			sum = digitA + digitB + carry;

			tempSum = String(sum % 10) + tempSum;
			carry = Math.floor(sum / 10);

			(i >= 0) ? i-- : i;
			(j >= 0) ? j-- : j;
		}

		start = 0;
		end = k - 1;
		while (start < end) {
			temp = tempSum[start];
			tempSum[start] = tempSum[end];
			tempSum[end] = temp;
			start++;
			end--;
		}

		globSum = tempSum;
	}

	/** return add and globSum getter (getSum) in obj **/
	return { add, getSum: () => globSum };
}

function fibonacci(n) {
	/** create closure instance **/
	let addClosure = createAddClosure();

	/** extract add **/
	let add = addClosure.add
	
	let a = '1';
	let b = '2';

	process.stdout.write(`${a}, ${b}`);

	let i = 3;
	while (i <= n) {
		add(a, b);
		process.stdout.write(`, ${addClosure.getSum()}`); /** call getter **/
		a = b;
		b = addClosure.getSum();
		i++;
	}
}

function driver() {
	const n = 98;
	fibonacci(n);
	process.stdout.write('\n');
}

driver()

File: 0x02-functions_nested_loops/fib.c
#include <stdio.h>
#include <string.h>

void add(char *a, char *b, char *result) {
    int carry = 0;
    int len_a = strlen(a);
    int len_b = strlen(b);
    int i = len_a - 1;
    int j = len_b - 1;
    int k = 0;

    a = "1" | b = "2" | result = "3\0" | len_a = 1 | len_b = 1 | i = -1 | j = -1 | k = 1 | carry = 0
    digit_a = 1 | digit_b = 2 | carry = 0 | sum = 3 | result[0] = '3'


    char z = '0'; 
    while (i >= 0 || j >= 0 || carry) {
        int digit_a = (i >= 0) ? (a[i] - '0') : 0;
        int digit_b = (j >= 0) ? (b[j] - '0') : 0;
        int sum = digit_a + digit_b + carry;

        result[k++] = (sum % 10) + '0';
        carry = sum / 10;

        if (i >= 0) i--;
        if (j >= 0) j--;
    }

    result[k] = '\0';

    // Reverse the result string
    int start = 0, end = k - 1;
    while (start < end) {
        char temp = result[start];
        result[start] = result[end];
        result[end] = temp;
        start++;
        end--;
    }
}

void fibonacci(int n) {
    char a[1000] = "1"; // Initialize a with 1
    char b[1000] = "2"; // Initialize b with 2
    char sum[1000];

    printf("%s", a);
    printf(", %s", b);

    for (int i = 3; i <= n; i++) {
        add(a, b, sum);
        printf(", %s", sum);

        strcpy(a, b); // Update a to b
        strcpy(b, sum); // Update b to sum
    }
}

int main(void) {
    int n = 98;
    printf("Fibonacci sequence up to %dth number:\n", n);
    fibonacci(n);
    printf("\n");
    return 0;
}

